### PR TITLE
COL-413 Script for Node REPL with preloaded Collabosphere environment

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.ets.berkeley.edu/jira/browse/COL-413

I've found this handy enough during local development that I decided to submit it. Rails devs will recognize it as a sketchy gesture toward a Rails console-like environment.

Example usage:

```
 % node scripts/col-repl
Loading Collabosphere environment...
[2016-07-21T19:29:48.883Z]  INFO: col-core/email/43411 on MacBook-Pro-2.local: No image manifest found in the document root. Using development assets (documentRoot=/var/www/html)
[2016-07-21T19:29:48.900Z]  INFO: col-activities/43411 on MacBook-Pro-2.local: Scheduling daily notifications for 2016-07-22T08:00:00-07:00 (in 70211998 ms)
Collabosphere environment loaded. Enjoy!
col-repl> CourseAPI.getCourses(function(err, courses) { log.info(courses[0].dataValues); });
undefined
[2016-07-21T19:29:59.723Z]  INFO: Intermediate Hittite/43411 on MacBook-Pro-2.local:  (id=285, canvas_course_id=1, enable_upload=true, assetlibrary_url=http://localhost:3100/courses/1/external_tools/4, whiteboards_url=http://localhost:3100/courses/1/external_tools/5, active=true, created_at=2016-05-23T20:25:34.170Z, updated_at=2016-05-23T20:26:37.183Z, canvas_api_domain=localhost:3100)
    
    --
    canvas: {
      "canvas_api_domain": "localhost:3100",
      "use_https": false,
      "name": "Localhost",
      "logo": "none",
      "created_at": "2016-05-16T23:17:55.076Z",
      "updated_at": "2016-05-16T23:17:55.076Z"
    }

col-repl> ActivitiesAPI.getActivities({course: {id: 285}}, null, null, null, function(err, activities) { log.info(activities.length + ' activities found'); });
undefined
[2016-07-21T19:30:16.058Z]  INFO: col-repl/43411 on MacBook-Pro-2.local: 33 activities found
col-repl> 
```

### CODE REVIEW

File: scripts/col-repl.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * REGENTS SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
+ * SOFTWARE AND ACCOMPANYING DOCUMENTATION, IF ANY, PROVIDED HEREUNDER IS PROVIDED
+ * "AS IS". REGENTS HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, UPDATES,
+ * ENHANCEMENTS, OR MODIFICATIONS.
+ */
+
+var _ = require('lodash');
+var DB = require('col-core/lib/db');
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: These should be separated by an additional line break as `_` is an external dependency and `DB` is an internal dependency

File: scripts/col-repl.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * Load Collabosphere environment and start REPL
+ */
+console.log('Loading Collabosphere environment...');
+
+var init = function() {
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: Can you add some JS docs about what REPL does ?

File: scripts/col-repl.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    
+  // Connect to the Collabosphere database
+  DB.init(function() {
+
+    // Require additional modules and utilities, assigning them to properties 
+    // to be exported to the REPL.
+
+    var exports = {
+      // External modules
+      'async': require('async'),
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: How do you determine which external modules are loaded and which aren't ?

File: scripts/col-repl.js [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      'config': require('config'),
+      'Joi': require('joi'),
+      // Node's REPL treats _ as a special variable, returning the result of the last 
+      // expression, so we can't assign to it.
+      'lodash': _,
+      'moment': require('moment-timezone'),
+      'Sequelize': require('sequelize'),
+
+      // Collabosphere modules
+      'ActivitiesAPI': require('col-activities'),
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: Same here. How do we decide what to include (e.g., the Whiteboards API is not included)

File: scripts/col-repl.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      'log': require('col-core/lib/logger')('col-repl'),
+      'RestAPI': require('col-rest'),
+      'UserConstants': require('col-users/lib/constants'),
+      'UsersAPI': require('col-users')
+    };
+
+    console.log('Collabosphere environment loaded. Enjoy!');
+
+    // Start the REPL and make exports available.
+
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: Remove line break

File: scripts/col-repl.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    var repl = require('repl').start({prompt: 'col-repl> '});
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: Can we do the `require('repl')` at the top with the other external includes?

**Comment by nicolaasmatthijs** [comment]
This seems useful. I'll definitely give it a try.

**Comment by nicolaasmatthijs** [comment]
Finished a round of review. Back to @paulkerschen for follow-up.

File: scripts/col-repl.js (hunk 2 above)
Comment (paulkerschen) [inline scripts/col-repl.js]: I can.

File: scripts/col-repl.js (hunk 3 above)
Comment (paulkerschen) [inline scripts/col-repl.js]: It's arbitrary; just a set of frequently used modules I found myself using in the console. Any other heuristic you'd recommend?

File: scripts/col-repl.js (hunk 4 above)
Comment (paulkerschen) [inline scripts/col-repl.js]: WhiteboardsAPI unfortunately requires a running server on include, at least as it presently exists. I can add a comment to that effect.

Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: That might be useful. Thanks!

**Comment by paulkerschen** [comment]
Review addressed, back to @nicolaasmatthijs.

File: scripts/col-repl.js [hunk 7]
```diff
@@ -28,11 +28,12 @@ var repl = require('repl');
...
 var DB = require('col-core/lib/db');
 
-/**
- * Load Collabosphere environment and start REPL
- */
 console.log('Loading Collabosphere environment...');
 
+/**
+ * Load Collabosphere environment and start an instance of Node's REPL (read-eval-print loop), 
+ an interactive interpreter that evaluates input line by line and displays evaluation results.
```
Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: This needs a `*` before it.

Comment (nicolaasmatthijs) [inline scripts/col-repl.js]: Also, can you add an `@see` at the bottom of this block pointing to the Node REPL repo ?

**Comment by nicolaasmatthijs** [comment]
Back to @paulkerschen for final review items

**Comment by paulkerschen** [comment]
Review addressed; back to @nicolaasmatthijs.

**Comment by nicolaasmatthijs** [comment]
Looks good. Merged.

**Pull request merged**